File: LeetCode-Python/030.py
class Solution:
    def findSubstring(self, s, words):
        """
        :type s: str
        :type words: List[str]
        :rtype: List[int]
        """
        dic = {}
        for word in words:
            if word not in dic:
                dic[word] = 1
            else:
                dic[word] += 1
        step = len(words[0])
        length = len(s)
        num = len(words)
        result = []
        for i in range(length-step*num+1):
            d = {}
            right = True
            for j in range(num):
                stemp = s[i+j*step:i+(j+1)*step]
                if stemp not in dic:
                    right = False
                    break
                elif stemp not in d:
                    d[stemp] = 1
                else:
                    d[stemp] += 1
                    if d[stemp] > dic[stemp]:
                        right = False
                        break
            if right == True:
                result.append(i)
        return result

File: LeetCode-C++/030.cpp
class Solution {
public:
	vector<int> findSubstring(string s, vector<string>& words) {
		int len = s.size();
		int m = words.size();
		int step = words[0].size();
		unordered_map<string, int> wordMap;
		for (int i = 0; i < m; i++) {
			wordMap[words[i]] += 1;
		}
		int stop = len - m*step + 1;
		vector<int> result;
		result.reserve(stop);
		for (int i = 0; i < stop; i++) {
			unordered_map<string, int> tempMap = wordMap;
			int j;
			for (j = 0; j < m; j++) {
				string temp = "";
				int start = i + j*step;
				int end = i + (j + 1)*step;
				for (int k = start; k < end; k++) {
					temp += s[k];
				}
				if (tempMap[temp] != 0) {
					tempMap[temp]--;
				}
				else {
					break;
				}
			}
			if (j == m) {
				result.push_back(i);
			}
		}
		return result;
	}
};